src/QuadForm/Quad/ZGenus.jl

```diff
@@ -510,60 +518,73 @@ Return the genus of the orthogonal direct sum of `G1` and `G2`.
+#
+###############################################################################
 
 @doc Markdown.doc"""
-    genera(sig_pair::Vector{Int}, determinant::Union{Int,fmpz};
-           max_scale=nothing, even=false) -> Vector{ZGenus}
+    genera(sig_pair::Vector{Int}, determinant::RationalUnion;
+           min_scale::RationalUnion = min(one(QQ), QQ(abs(determinant))),
+           max_scale::RationalUnion = max(one(QQ), QQ(abs(determinant))),
+           even=false)                                         -> Vector{ZGenus}
 
-Return a list of all genera with the given conditions.
+Return a list of all genera with the given conditions. Genera of non-integral
+$\mathbb Z$-lattices are also supported.
 
 # Arguments
 - `sig_pair`: a pair of non-negative integers giving the signature
-- `determinant`: an integer; the sign is ignored
-- `max_scale`: (default: `nothing`) an integer; the maximum scale of a
-      jordan block
-- `even`: boolean (default: `false`)
-"""
-function genera(sig_pair::Tuple{Int,Int}, determinant::Union{Int,fmpz};
-                max_scale=nothing, even=false)
-  @req all(s >= 0 for s in sig_pair) "the signature vector must be a pair of non negative integers."
-  if max_scale == nothing
-    _max_scale = determinant
-  else
-    _max_scale = FlintZZ(max_scale)
-  end
+- `determinant`: a rational number; the sign is ignored
+- `min_scale`: a rational number; product of minimum scales for the jordan blocks
+  corresponding to the primes dividing the numerator or the denominator of
+  `determinant` (default: `min(one(QQ), QQ(abs(determinant)))`)
+- `max_scale`: a rational number; product of maximum scales for the jordan blocks
+  corresponding to the primes dividing the numerator or the denominator of
+  `determinant` (default: `min(one(QQ), QQ(abs(determinant)))`)
+- `even`: boolean; if set to true, return only the even genera (default: `false`)
+"""
+function genera(sig_pair::Tuple{Int,Int}, determinant::RationalUnion;
+                min_scale::RationalUnion = min(one(QQ), QQ(abs(determinant))),
+                max_scale::RationalUnion = max(one(QQ), QQ(abs(determinant))),
+                even=false)
+  @req all(s >= 0 for s in sig_pair) "The signature vector must be a pair of non negative integers."
+  determinant = QQ(determinant)
+  denominator(determinant) != 1 && even && return ZGenus[]
+  @req min_scale > 0 "Minimal scale must be a positive integer"
+  min_scale = QQ(min_scale)
+  @req max_scale > 0 "Maximal scale must be a positive integer"
+  max_scale = QQ(max_scale)
   rank = sig_pair[1] + sig_pair[2]
   out = ZGenus[]
   local_symbols = Vector{ZpGenus}[]
   # every global genus has a 2-adic symbol
-  if mod(determinant, 2) == 1
-    push!(local_symbols, _local_genera(ZZ(2), rank, 0, 0, even))
+  if mod(numerator(determinant)*denominator(determinant), 2) == 1
+    push!(local_symbols, _local_genera(2, rank, 0, 0, 0, even))
   end
   # collect the p-adic symbols
-  for p in prime_divisors(determinant)
```

Comment:
I just modified it so that we take in account also the primes for which the valuations of `min_scale` and `max_scale` are non trivial... As for the bounds on the scales, if the determinant is integral and `min_scale` is not specified, we get the same as before, if the determinant is strictly rational, then we adapt by imposing the maximal scale to be 1 if not specified (it is a choice to be consistent with what is happening on the other side). Of course we can change this "convention" but I guess it makes sense in general.